src/codegen/operator/index_scan_translator.cpp

```diff
@@ -0,0 +1,343 @@
+      }
+
+      // Iterate over the batch using a scalar loop
+      batch.Iterate(codegen, [&](RowBatch::Row &row) {
+        // Evaluate the predicate to determine row validity
+        codegen::Value valid_row = row.DeriveValue(codegen, *predicate);
+
+        // Reify the boolean value since it may be NULL
+        PL_ASSERT(valid_row.GetType().GetSqlType() ==
+                  type::Boolean::Instance());
+        llvm::Value *bool_val =
+            type::Boolean::Instance().Reify(codegen, valid_row);
+
+        // Set the validity of the row
+        row.SetValidity(codegen, bool_val);
+      });
+    }
+
+    // construct the final row batch
+    // one tuple per row batch
+    RowBatch final_batch{this->GetCompilationContext(), tile_group_id,
+                         codegen.Const32(0), codegen.Const32(1), sel_vec, true};
+
+    std::vector<TableScanTranslator::AttributeAccess> final_attribute_accesses;
+    std::vector<const planner::AttributeInfo *> final_ais;
+    index_scan_.GetAttributes(final_ais);
+    std::vector<oid_t> output_col_ids;
+    if (index_scan_.GetColumnIds().size() != 0) {
+      output_col_ids = index_scan_.GetColumnIds();
+    } else {
+      output_col_ids.resize(table.GetSchema()->GetColumnCount());
+      std::iota(output_col_ids.begin(), output_col_ids.end(), 0);
+    }
+    //    const auto &output_col_ids = index_scan_.GetColumnIds();
+
+    for (oid_t col_idx = 0; col_idx < output_col_ids.size(); col_idx++) {
+      final_attribute_accesses.emplace_back(tile_group_access,
+                                            final_ais[output_col_ids[col_idx]]);
+    }
+    for (oid_t col_idx = 0; col_idx < output_col_ids.size(); col_idx++) {
+      auto *attribute = final_ais[output_col_ids[col_idx]];
+      final_batch.AddAttribute(attribute, &final_attribute_accesses[col_idx]);
+    }
+
+    ConsumerContext context{this->GetCompilationContext(), this->GetPipeline()};
+    context.Consume(final_batch);
+
+    // Move to next tuple in the index scan result
+    result_iter = codegen->CreateAdd(result_iter, codegen.Const64(1));
+    loop.LoopEnd(codegen->CreateICmpULT(result_iter, result_size),
+                 {result_iter});
+  }
+
+  // free the memory allocated for the index scan iterator
+  codegen.Call(RuntimeFunctionsProxy::DeleteIterator, {iterator_ptr});
+}
+
+// Get the name of this scan
+std::string IndexScanTranslator::GetName() const {
+  std::string name = "Scan('" + GetIndex().GetName() + "'";
+  auto *predicate = GetIndexScanPlan().GetPredicate();
+  if (predicate != nullptr && predicate->IsSIMDable()) {
+    name.append(", ").append(std::to_string(Vector::kDefaultVectorSize));
+  }
+  name.append(")");
+  return name;
+}
+
+// Index accessor
+const index::Index &IndexScanTranslator::GetIndex() const {
+  return dynamic_cast<index::Index &>(*index_scan_.GetIndex().get());
+}
+
+void IndexScanTranslator::UpdateTupleWithParameterCache(
```

Comment:
Name this as `SetPredicate` or something that shows what is doing in the perspective of IndexScan.  `TupleWithParameterCache` is an implementation detail.  Additionally, move the predicate check outside the function for readability.